### PR TITLE
added code and tests for mismatched arrays, CRS, and Affine xform in …

### PR DESCRIPTION
**Pull request by joemcglinchy**
…stack function.

### CODE REVIEW

**Comment by joemcglinchy** [comment]
Note that there was one failing test, but unrelated to my code (not sure what was going on!). See below.

```
___________________________________________________________________________________________ test_urls_are_valid ___________________________________________________________________________________________

    @skip_on_ci
    @pytest.mark.vcr()
    def test_urls_are_valid():
        """ Test responses for each dataset to ensure valid URLs. """
        for key in eio.DATA_URLS:
            dataset = eio.DATA_URLS[key]
            if not isinstance(dataset, list):
                dataset = [dataset]
            for url, name, kind in dataset:
>               r = requests.get("http://www.example.com")

earthpy\tests\test_io.py:109:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\requests\api.py:75: in get
    return request('get', url, params=params, **kwargs)
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\requests\api.py:60: in request
    return session.request(method=method, url=url, **kwargs)
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\requests\sessions.py:533: in request
    resp = self.send(prep, **send_kwargs)
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\requests\sessions.py:646: in send
    r = adapter.send(request, **kwargs)
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\requests\adapters.py:449: in send
    timeout=timeout
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\urllib3\connectionpool.py:603: in urlopen
    chunked=chunked)
c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\urllib3\connectionpool.py:379: in _make_request
    httplib_response = conn.getresponse(buffering=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <vcr.patch.VCRRequestsHTTPConnectionC:\Projects\earth_py_support\joe_fork_github\earthpy\earthpy\tests\cassettes\test_urls_are_valid.yaml object at 0x0000024C01682710>, _ = False
kwargs = {'buffering': True}

    def getresponse(self, _=False, **kwargs):
        '''Retrieve the response'''
        # Check to see if the cassette has a response for this request. If so,
        # then return it
        if self.cassette.can_play_response_for(self._vcr_request):
            log.info(
                "Playing response for {} from cassette".format(
                    self._vcr_request
                )
            )
            response = self.cassette.play_response(self._vcr_request)
            return VCRHTTPResponse(response)
        else:
            if self.cassette.write_protected and self.cassette.filter_request(
                self._vcr_request
            ):
                raise CannotOverwriteExistingCassetteException(
                    "No match for the request (%r) was found. "
                    "Can't overwrite existing cassette (%r) in "
                    "your current record mode (%r)."
                    % (self._vcr_request, self.cassette._path,
>                      self.cassette.record_mode)
                )
E               vcr.errors.CannotOverwriteExistingCassetteException: No match for the request (<Request (GET) http://www.example.com/>) was found. Can't overwrite existing cassette ('C:\\Projects\\earth_py_support\\joe_fork_github\\earthpy\\earthpy\\tests\\cassettes\\test_urls_are_valid.yaml') in your current record mode ('once').

c:\software\anaconda3\envs\earthpy-multiclip\lib\site-packages\vcr\stubs\__init__.py:237: CannotOverwriteExistingCassetteException
================================================================================== 1 failed, 139 passed in 4.85 seconds ================
```

**Comment by joemcglinchy** [comment]
@lwasser @mbjoseph ready for review at your leisure!

**Comment by lwasser** [comment]
thank you @joemcglinchy a few small items 
1. everything looks good but just make sure you've updated from master recently!
2. can you please update the [changelog](https://raw.githubusercontent.com/earthlab/earthpy/master/CHANGELOG.md) with a statement about what you did and tie it to the ISSUE number:
most recent example below:
```
## [unreleased]
* Add vignette for using the plot_bands() function (@nkorinek, #315)
```
your add would be a bullet point ABOVE that one. again # referencing the issue number ! thank you!!

**Comment by lwasser** [comment]
addresses issue: https://github.com/earthlab/earthpy/issues/334

**Comment by joemcglinchy** [comment]
@lwasser is item 1 above to try and fix the errors I am getting in `test_urls_are_valid`? I'm not sure why that test is failing for me, but the tests for the updated stack functionality are all passing and there haven't been any updates to master since I merged and created the feature branch.

Item 2 is taken care of in the most recent commit.

**Comment by lwasser** [comment]
@joemcglinchy all of the tests seem to be passing in CI and when i grab this PR and run it locally (building on top of master). is it just failing for you locally? i think we had made some changes to skip tests that required an internet connection. but please clarify when it's failing. it seems ok to me!

**Comment by joemcglinchy** [comment]
the `test_urls_are_valid` is the only test that is failing for me when I run `pytest`. See the traceback from pytests above. https://github.com/earthlab/earthpy/pull/344#issuecomment-498811220

**Comment by mbjoseph** [comment]
@joemcglinchy you may need to delete your local cassettes directory and re-run the tests.

**Comment by lwasser** [comment]
ahhhh i wondered if i was missing something. thank you @mbjoseph 
my cassettes had been cleaned out and are recent. 

Screen Shot 2019-06-05 at 1 00 34 PM

**Comment by joemcglinchy** [comment]
woo hoo that was it thanks @mbjoseph, tape deck is operable.

File: earthpy/tests/test_stack.py [hunk 1]
```diff
@@ -18,6 +18,45 @@ def in_paths(basic_image_tif):
...
 
 
+@pytest.fixture
+def in_paths_mismatch(basic_image_tif, basic_image_tif_2):
+    """ Input file paths for tifs of different sizes to stack. """
+    return [basic_image_tif] * 4 + [basic_image_tif_2]
+
+
+@pytest.fixture
+def in_paths_CRS_mismatch(basic_image_tif, basic_image_tif_CRS):
```
Comment (lwasser) [inline earthpy/tests/test_stack.py]: SHould this be 4 images with different CRS' to stack?

Comment (lwasser) [reply earthpy/tests/test_stack.py]: @nkorinek just a note that you may be able to use these fixtures in your crop_all tests.

File: earthpy/spatial.py [hunk 2]
```diff
@@ -192,7 +192,29 @@ def stack(band_paths, out_path="", nodata=None):
...
+            )
+
+        if not len(set(dest_aff)) == 1:
+            raise ValueError(
+                "Please ensure all source rasters have same affine transform."
+            )
+
+        if not len(set(dest_shps)) == 1:
+            raise ValueError(
+                "Please ensure all source rasters have same rows and columns."
```
Comment (lwasser) [inline earthpy/spatial.py]: How about `Please ensure all source rasters have the same shape (nrows, ncols)`

File: earthpy/tests/test_stack.py [hunk 3]
```diff
@@ -18,6 +18,45 @@ def in_paths(basic_image_tif):
     return [basic_image_tif] * 4
 
 
+@pytest.fixture
+def in_paths_mismatch(basic_image_tif, basic_image_tif_2):
+    """ Input file paths for tifs of different sizes to stack. """
```
Comment (lwasser) [inline earthpy/tests/test_stack.py]: Is this different sizes or different dimensions? ie the shape is different?

File: earthpy/tests/test_stack.py [hunk 4]
```diff
@@ -18,6 +18,45 @@ def in_paths(basic_image_tif):
...
+
+@pytest.fixture
+def in_paths_CRS_mismatch(basic_image_tif, basic_image_tif_CRS):
+    """ Input file paths for tifs of different sizes to stack. """
+    return [basic_image_tif] * 4 + [basic_image_tif_CRS]
+
+
+@pytest.fixture
+def in_paths_Affine_mismatch(basic_image_tif, basic_image_tif_Affine):
+    """ Input file paths for tifs of different sizes to stack. """
```
Comment (lwasser) [inline earthpy/tests/test_stack.py]: dimensions? or dimensions (nrows, ncols)

File: earthpy/tests/test_stack.py [hunk 5]
```diff
@@ -18,6 +18,45 @@ def in_paths(basic_image_tif):
...
+    return [basic_image_tif] * 4 + [basic_image_tif_Affine]
+
+
+def test_stack_array_size_mismatch(in_paths_mismatch):
+    """ Test for error raised when array sizes are not all equal. """
```
Comment (lwasser) [inline earthpy/tests/test_stack.py]: array dimensions   (that just be a bit more explicit)

**Comment by lwasser** [comment]
@mbjoseph can you have a look at this pr? this error below is not related to this PR but i'mm getting it when testing this pr and my cassettes have been cleared out. any thoughts?
```
___________________________________ test_arbitrary_url_zip_download ____________________________________

eld = Available Datasets: ['california-rim-fire', 'co-flood-extras', 'cold-springs-fire', 'cold-springs-modis-h5', 'colorado...valid-data-type', 'little-text-file', 'little-zip-file', 'ndvi-automation', 'spatial-vector-lidar', 'vignette-landsat']

    @skip_on_ci
    @pytest.mark.vcr()
    def test_arbitrary_url_zip_download(eld):
        """ Verify that aribitrary URLs work for zip file downloads. """
        path = eld.get_data(
>           url="https://www2.census.gov/geo/tiger/GENZ2016/shp/cb_2016_us_nation_20m.zip"
        )

earthpy/tests/test_io.py:193: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
earthpy/io.py:221: in get_data
    verbose=verbose,
earthpy/io.py:269: in _download
    self._download_and_extract(path, r, kind, verbose)
earthpy/io.py:297: in _download_and_extract
    archive = zipfile.ZipFile(file_like_object)
/anaconda3/envs/earthpy-dev/lib/python3.6/zipfile.py:1131: in __init__
    self._RealGetContents()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <zipfile.ZipFile [closed]>

    def _RealGetContents(self):
        """Read in the table of contents for the ZIP file."""
        fp = self.fp
        try:
            endrec = _EndRecData(fp)
        except OSError:
            raise BadZipFile("File is not a zip file")
        if not endrec:
>           raise BadZipFile("File is not a zip file")
E           zipfile.BadZipFile: File is not a zip file

/anaconda3/envs/earthpy-dev/lib/python3.6/zipfile.py:1198: BadZipFile
----------------------------------------- Captured stdout call -----------------------------------------
Downloading from https://www2.census.gov/geo/tiger/GENZ2016/shp/cb_2016_us_nation_20m.zip
```

File: earthpy/tests/test_stack.py (hunk 1 above)
Comment (joemcglinchy) [reply earthpy/tests/test_stack.py]: @lwasser I don't think we need 4 different CRS, just 2 different ones at a minimum. The check is for the `set` of CRS present in the source list to be == 1. Having 2 CRS here supports that test.

File: earthpy/tests/test_stack.py (hunk 3 above)
Comment (joemcglinchy) [reply earthpy/tests/test_stack.py]: Correct, the shapes of basic_image_tif and basic_image_tif_2 are different (10x10 and 20x20, respectively).

File: earthpy/tests/test_stack.py (hunk 4 above)
Comment (joemcglinchy) [reply earthpy/tests/test_stack.py]: copy pasta. I'll update :)

File: earthpy/tests/test_stack.py (hunk 5 above)
Comment (joemcglinchy) [reply earthpy/tests/test_stack.py]: sure thing

File: earthpy/tests/test_stack.py (hunk 4 above)
Comment (lwasser) [reply earthpy/tests/test_stack.py]: copypasta-all-the-spaghetti-code

File: earthpy/spatial.py (hunk 2 above)
Comment (joemcglinchy) [reply earthpy/spatial.py]: done

**Comment by mbjoseph** [comment]
@lwasser I cannot reproduce that error - my best guess is that maybe the census.gov URL that the test hits was not available briefly, but it's hard to know without being able to reproduce/dig into it.

**Comment by lwasser** [comment]
ok. well if you can't reproduce it, i'll assume it's something like that because CI is happy and it works for you as well. my internet was a bit pokie yesterday in the office to perhaps that was the issue... let's assume it's ok then :)

**Pull request merged**